### PR TITLE
add metrics and log for ticlient

### PR DESCRIPTION
**Pull request by wentaoxu**
as title says

### CODE REVIEW

**Comment by wentaoxu** [comment]
PTAL @disksing @zhangjinpeng1987 @AndreMouche @coocood @siddontang

**Comment by shenli** [comment]
@wentaoxu Please fix the CI.

File: store/tikv/client.go [hunk 1]
```diff
@@ -100,18 +100,20 @@ func (a *connArray) Init(addr string, security config.Security) error {
...
-			grpc.WithTimeout(dialTimeout),
 			grpc.WithInitialWindowSize(grpcInitialWindowSize),
 			grpc.WithInitialConnWindowSize(grpcInitialConnWindowSize),
 			grpc.WithUnaryInterceptor(unaryInterceptor),
 			grpc.WithStreamInterceptor(streamInterceptor),
 			grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(MaxCallMsgSize)),
 			grpc.WithDefaultCallOptions(grpc.MaxCallSendMsgSize(MaxSendMsgSize)),
 		)
-
+		cancel()
```
Comment (shenli) [inline store/tikv/client.go]: Why call cancel() here?

File: store/tikv/gcworker/gc_worker.go [hunk 2]
```diff
@@ -104,7 +104,8 @@ const (
 	gcDefaultLifeTime        = time.Minute * 10
 	gcSafePointKey           = "tikv_gc_safe_point"
 	gcSafePointCacheInterval = tikv.GcSafePointCacheInterval
-	gcScanLockLimit          = 1000
+	// we don't want gc to sweep out the cached info belong to other processes, like coprocessor
```
Comment (shenli) [inline store/tikv/gcworker/gc_worker.go]: Why gc will sweep out the cached info?

File: store/tikv/lock_resolver.go [hunk 3]
```diff
@@ -27,7 +27,8 @@ import (
 	goctx "golang.org/x/net/context"
 )
 
-const resolvedCacheSize = 512
+// ResolvedCacheSize max number of cached txn status
```
Comment (shenli) [inline store/tikv/lock_resolver.go]: end with `.`

File: store/tikv/safepoint.go [hunk 4]
```diff
@@ -27,6 +27,9 @@ import (
 
 // Safe point constants.
 const (
+	// this is almost the same as 'tikv_gc_safe_point' in the table 'mysql.tidb',
```
Comment (shenli) [inline store/tikv/safepoint.go]: Capitalize the first letter.

File: store/tikv/gcworker/gc_worker.go (hunk 2 above)
Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: gc may need much transaction info even if in one scan-lock request, like 1000 or more, so other transaction info will be eliminated.

File: store/tikv/lock_resolver.go (hunk 3 above)
Comment (wentaoxu) [reply store/tikv/lock_resolver.go]: ok

File: store/tikv/safepoint.go (hunk 4 above)
Comment (wentaoxu) [reply store/tikv/safepoint.go]: ok

File: store/tikv/client.go (hunk 1 above)
Comment (wentaoxu) [reply store/tikv/client.go]: I think it is safe and to call `cancel` here, to release the resource as quick as possible.

Comment (tiancaiamao) [reply store/tikv/client.go]: If grpc handle cancel signal right (i.e, never store the ctx in its inner state), this confirms to the Go idiom to release resource.
But I'm not sure of grpc implementation. So is there any example of this kind of usage on their canonical website of document? @wentaoxu

File: store/tikv/gcworker/gc_worker.go (hunk 2 above)
Comment (tiancaiamao) [reply store/tikv/gcworker/gc_worker.go]: Capital the first letter and add a dot at the end of comment sentence.

File: store/tikv/gcworker/misc.go [hunk 5]
```diff
@@ -50,11 +50,20 @@ var (
 			Name:      "gc_failure",
 			Help:      "Counter of gc job failure.",
 		}, []string{"type"})
+
+	gcRegionTooMuchLocksCounter = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: "tidb",
+			Subsystem: "tikvclient",
+			Name:      "gc_region_too_much_locks",
+			Help:      "Counter of gc scan lock req more than once in the same region.",
```
Comment (tiancaiamao) [inline store/tikv/gcworker/misc.go]: req means require or request?

File: store/tikv/kv.go [hunk 6]
```diff
@@ -241,7 +241,7 @@ func (s *tikvStore) runSafePointChecker() {
 				d = gcSafePointUpdateInterval
 			} else {
 				loadSafepointCounter.WithLabelValues("fail").Inc()
-				log.Errorf("[tikv] fail to load safepoint: %v", err)
+				log.Errorf("[check visibility] fail to load safepoint from pd: %v", err)
```
Comment (tiancaiamao) [inline store/tikv/kv.go]: Maybe this gives more information?
[tikv] check visibility fail to ...

**Comment by wentaoxu** [comment]
PTAL @shenli

File: store/tikv/metrics.go [hunk 7]
```diff
@@ -69,15 +69,6 @@ var (
 			Buckets:   prometheus.ExponentialBuckets(0.0005, 2, 20),
 		})
 
-	connPoolHistogram = prometheus.NewHistogramVec(
```
Comment (tiancaiamao) [inline store/tikv/metrics.go]: This is removed? why?

File: store/tikv/region_cache.go [hunk 8]
```diff
@@ -545,6 +545,11 @@ type RegionVerID struct {
 	ver     uint64
 }
 
+// GetID returns the id of the region
+func (r *RegionVerID) GetID() uint64 {
```
Comment (tiancaiamao) [inline store/tikv/region_cache.go]: Is this thread safe and does it need to be? @disksing

File: store/tikv/gcworker/gc_worker.go (hunk 2 above)
Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: ok

File: store/tikv/gcworker/misc.go (hunk 5 above)
Comment (wentaoxu) [reply store/tikv/gcworker/misc.go]: It means request, but I think I should make it clear here.

File: store/tikv/region_cache.go (hunk 8 above)
Comment (wentaoxu) [reply store/tikv/region_cache.go]: I think it is safe here because it will never change. PTAL @disksing

File: store/tikv/metrics.go (hunk 7 above)
Comment (wentaoxu) [reply store/tikv/metrics.go]: this is useless after we use grpc instead of tcp, dead code for long time.

File: store/tikv/kv.go (hunk 6 above)
Comment (wentaoxu) [reply store/tikv/kv.go]: The message comes from pd, not tikv, so I remove the flag. I think `check visibility` is useless either.^^

File: store/tikv/gcworker/gc_worker.go [hunk 9]
```diff
@@ -104,7 +104,8 @@ const (
 	gcDefaultLifeTime        = time.Minute * 10
 	gcSafePointKey           = "tikv_gc_safe_point"
 	gcSafePointCacheInterval = tikv.GcSafePointCacheInterval
-	gcScanLockLimit          = 1000
+	// We don't want gc to sweep out the cached info belong to other processes, like coprocessor.
+	gcScanLockLimit = tikv.ResolvedCacheSize / 2
```
Comment (coocood) [inline store/tikv/gcworker/gc_worker.go]: This is an irrelevant change, I think we should change it in another PR.

File: store/tikv/lock_resolver.go [hunk 10]
```diff
@@ -28,7 +28,8 @@ import (
 	goctx "golang.org/x/net/context"
 )
 
-const resolvedCacheSize = 512
+// ResolvedCacheSize max number of cached txn status.
+const ResolvedCacheSize = 2048
```
Comment (coocood) [inline store/tikv/lock_resolver.go]: ditto

File: store/tikv/gcworker/gc_worker.go (hunk 9 above)
Comment (wentaoxu) [reply store/tikv/gcworker/gc_worker.go]: ok, It should be.

File: store/tikv/lock_resolver.go (hunk 10 above)
Comment (wentaoxu) [reply store/tikv/lock_resolver.go]: ok

**Comment by coocood** [comment]
LGTM

File: store/tikv/lock_resolver.go [hunk 11]
```diff
@@ -28,7 +28,8 @@ import (
 	goctx "golang.org/x/net/context"
 )
 
-const resolvedCacheSize = 512
+// ResolvedCacheSize max number of cached txn status.
+const ResolvedCacheSize = 512
```
Comment (tiancaiamao) [inline store/tikv/lock_resolver.go]: Is it necessary to make this variable public when no where else use it?

File: store/tikv/region_cache.go (hunk 8 above)
Comment (tiancaiamao) [reply store/tikv/region_cache.go]: So `getID` is enough?

File: store/tikv/lock_resolver.go (hunk 11 above)
Comment (wentaoxu) [reply store/tikv/lock_resolver.go]: I have move it to another pr #5788.

File: store/tikv/region_cache.go (hunk 8 above)
Comment (disksing) [reply store/tikv/region_cache.go]: Both fine.

**Comment by tiancaiamao** [comment]
LGTM

File: store/tikv/gcworker/misc.go [hunk 12]
```diff
@@ -50,11 +50,20 @@ var (
 			Name:      "gc_failure",
 			Help:      "Counter of gc job failure.",
 		}, []string{"type"})
+
+	gcRegionTooMuchLocksCounter = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: "tidb",
+			Subsystem: "tikvclient",
+			Name:      "gc_region_too_much_locks",
```
Comment (disksing) [inline store/tikv/gcworker/misc.go]: too much locks -> too many locks.
the rest LGTM.

**Pull request merged**